File: src/xstate/modules/schemas/drizzle/system.ts
import { index, text, integer, primaryKey } from 'drizzle-orm/sqlite-core';
export const system_fields = {
  id: text().primaryKey(),
  tombstone: integer().default(0),
  status: text().default('Active'),
  version: integer().default(1),
  created_date: text(),
  created_time: text(),
  updated_date: text(),
  updated_time: text(),
};
export const fileRegex = new RegExp(/.(ts|js)$/, 'gi');

const getDefaultIndices = (table_name: string) => {
  return (table) => {
    return Object.keys(system_fields).reduce((acc, field) => {
      const index_name = `${table_name}_${field}_idx`;
      if (index_name.includes('_time')) return acc;
      return {
        ...acc,
        [index_name]: index(index_name).on(table[field]),
      };
    }, {});
  };
};

const getPrimaryKey = (field_keys: string[]) => {
  return (table) => {
    return {
      pk: primaryKey({
        columns: field_keys.map((key) => table[key]) as any,
      }),
    };
  };
};

export const getConfigDefaults = {
  byIndex: getDefaultIndices,
  byPrimaryKey: getPrimaryKey,
};

File: bin/store/src/schema/system.rs
#[macro_export]
macro_rules! system_fields {
    () => {
        id -> Text,
        categories -> Nullable<Array<Nullable<Text>>>,
        code -> Nullable<Text>,
        tombstone -> Nullable<Int4>,
        status -> Nullable<Text>,
        previous_status -> Nullable<Text>,
        version -> Nullable<Int4>,
        created_date -> Nullable<Text>,
        created_time -> Nullable<Text>,
        updated_date -> Nullable<Text>,
        updated_time -> Nullable<Text>,
        organization_id -> Nullable<Text>,
        created_by -> Nullable<Text>,
        updated_by -> Nullable<Text>,
        deleted_by -> Nullable<Text>,
        requested_by -> Nullable<Text>,
        tags -> Nullable<Array<Nullable<Text>>>,
    };
}